website/blog/2024-11-27-test-smarter-part-2.md

```diff
@@ -0,0 +1,123 @@
+
+*Note that we are constructing this guidance based on how we [structure data at dbt Labs.](https://docs.getdbt.com/best-practices/how-we-structure/1-guide-overview#guide-structure-overview)* You may use a different modeling approach—that’s okay! Translate our guidance to your data’s shape, and let us know in the comments section what modifications you made. 
+
+First, here’s our opinions on where specific tests should go:
+
+- Source tests should be fixable data quality concerns. (See the callout box below for what we mean by “fixable”)
+- Staging tests should be business-focused anomalies specific to individual tables, such as accepted ranges or ensuring sequential values. In addition to these tests, your staging layer should clean up any nulls, duplicates, or outliers that you can’t fix in your source system. You generally don’t need to test your cleanup efforts.
+- Intermediate and marts layer tests should be business-focused anomalies resulting specifically from joins or calculations.  You also may consider adding additional primary key and not null tests on columns where it’s especially important to protect the grain.
+
+## Where should tests go in your pipeline?
+
+![A horizontal, multicolored diagram that shows examples of where tests ought to be placed in a data pipeline.](/img/blog/2024-11-27-test-smarter-part-2/testing_pipeline.png)
+
+This diagram above outlines where you might put specific data tests in your pipeline. Let’s expand on it and discuss where each type of data quality issue should be tested. 
+
+### Sources
+
+Tests applied to your sources should indicate *fixable-at-the-source-system* issues. If your source tests flag source system issues that aren’t fixable, remove the test and mitigate the problem in your staging layer instead. 
+
+:::info
+💡 A note on what we mean by *fixable*: We consider a "fixable-at-the-source-system" issue to be something that:
+
+- You yourself can fix in the source system.
+- You know the right person to fix it and have a good enough relationship with them that you know you can *get it fixed.*
+
+You may have issues that can *technically* get fixed at the source, but it won't happen till the next planning cycle, or you need to develop better relationships to get the issue fixed, or something similar. This demands a more nuanced approach than we'll cover in this post. If you have thoughts on this type of situation, let us know!
+
+:::
+
+Here’s our recommendation for what tests belong on your sources. 
+
+- Source freshness: testing data freshness for sources that are critical to your pipelines.
+    - If any sources feed into any of the “top 3” [priority categories](https://docs.getdbt.com/blog/test-smarter-not-harder#how-to-prioritize-data-quality-concerns-in-your-pipeline) in our last post, put [`dbt source freshness`](https://docs.getdbt.com/docs/deploy/source-freshness) in your job execution commands. If your sources are stale, dbt Cloud will fail your job.
```

Comment:
I also don't know if these two bullets have to be so Cloud-y - this advice all holds regardless of how you orchestrate, so I'd probably focus on the outcome and then link out to the "here's how you do this in dbt cloud" for convenience, as opposed to explaining what the Cloud configs themselves should be.

Reply:
updating the post to be less cloud-y! i disagree though about still letting the job run. Part of the goal here is to be opinionated about errors/where to put tests and to have errors actually have consequences (like a pipeline failing, instead of just more noise with alarms). 

I've had opinions vary widely on this in the field. Some people feel the way you do--no need for the pipeline to fail if there's stale data, just send me a warning. 

But, for other industries (I've had this come up most frequently in energy & financial companies), data staleness is a 5-alarm fire and they _want_ the pipeline going down so they have very material motivation to fix it. 

I'm open to adding some sort of a caveat or softening of the guidance above here for sure....and continuing to talk about it 😆 but as much as possible, I want this post to be decisive about errors having consequences and informing action, and I want as few "oh just send me a warning" moments as possible. Does that make sense? Further thoughts @joellabes ?

Reply:
If you're seeing a lot of divergent opinions in the field then it might be worth saying "If X is important then do Y, but if Z is more important then do A". But it's an opinion piece so I'm good with whatever you choose.